design/feature-flags.md

```diff
@@ -0,0 +1,98 @@
+# "Future" Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is
+the correct default behavior for the CDK. The problem, of course, is that
+breaking changes are only allowed in major versions and those are rare.
+
+This document describes a proposal for a pattern/mechanism called "future
+flags". It will allow us to introduce breaking behavior which is disabled by
+default (so existing projects will not be affected) but enabled automatically
+for new projects created through `cdk init`.
+
+> The name "**future flags**" is a wordplay on "feature flags" (dah!) to
+> represent features that are "from the future" and indicate that these specific
+> feature flags are only disabled by default only until the next release, after
+> which they will either be removed _or_ turned into regular feature flags that
+> people can select if they so desire.
+>
+> Having said all that, I am okay if we feel this is unnecessary and we can call
+> these just "feature flags" and not have a semantic distinction between the
+> ones that are "temporary" and permanent.
+
+## Approach
+
+The basic idea is that new breaking behavior will always be disabled by default
+and only enabled when a certain CDK context parameter is set. If not enabled,
+the system will continue to behave exactly like it used to without breaking any
+existing projects.
+
+When we release a new major version of the AWS CDK, we will flip this behavior
```

Comment:
I think we have a difference of expectation in how often we'll release new major versions. I'm expecting this to be a rare event, as the health of our 3rd party ecosystem is going to depend on that (`peerDependency ^1.26.3` will stop working when we release `2.0.0`, and `peerDependency >= 1.26.3` is not safe).

So I would expect us to (be allowed to) break MV compatibility every year at best, probably less. I would therefore like a solution that does not accumulate too much cruft over the lifetime of a major release.

Reply:
Indeed, we don't want to release too many major versions. Still I don't think we'll accumulate too many features that must break the behavior between major releases. But I guess we will have to wait and see.

Reply:
Added a new under a new section for "future development"